### PR TITLE
Phase 2 deep review C-1/C-2/C-3: critical security + correctness fixes

### PR DESCRIPTION
**Pull request by raroche**
## Summary

Three **Critical** findings from the post-Phase-2 deep review, addressing untrusted-input robustness in attacker-reachable code paths. **Companion to [PR #13](https://github.com/raroche/NetPdf/pull/13)** (which addresses the 7 P1/P2 items from the earlier targeted review). Stacks on top of #13's branch so the two can merge in either order. After both land, the deep review's actionable critical / high-severity findings are cleared and Phase 3 can build on a hardened Phase 2.

10 net-new tests, all passing. **3239 unit / 3365 solution-wide** (+10 over the prior 3229/3355 baseline). AOT/JIT byte-parity gate green.

### Phase 2 deep review — 3 Critical fixes ([e29c085](https://github.com/raroche/NetPdf/commit/e29c085))

**C-1 [CRITICAL] — `SelectorCompiler` recursion depth limit**

The recursive-descent selector parser had no depth bound on `:is()` / `:not()` / `:has()` nesting. An attacker-supplied stylesheet with a 10k-deep nested selector would crash the host with `StackOverflowException` (uncatchable, brings down the process). Added `MaxRecursionDepth = 64` + a `_depth` counter tracked across the `ParseSelectorList` chain — mirrors the existing `VarSubstitution.MaxRecursionDepth` defense pattern. The guard fires as a catchable `SelectorParseException` which `CascadeResolver` wraps in `CSS-PARSE-WARNING-001` + skips the rule.

Note: forgiving sub-groups `:is()` / `:where()` per CSS Selectors L4 §3.7 correctly *swallow* the exception + drop the alternative — the guard manifests as a propagated exception only in strict contexts (`:not()` / `:has()`), which is spec-correct behavior. Tests verify both the propagation in strict context AND that reasonable nesting (5 levels) still compiles cleanly.

**C-2 [CRITICAL] — Diagnostic message sanitization**

The `CSS-PARSE-WARNING-001` emission in `CompileSelectorWithDiagnostics` interpolated the raw selector text + the exception reason verbatim into the message string:

```
\$\"Invalid selector \\\"{raw}\\\" — {ex.Reason}. Rule skipped.\"
```

Untrusted CSS can carry ANSI escape sequences, NUL / control chars, or extreme length. Emitting verbatim let a hostile stylesheet **inject control sequences into any downstream sink** that interprets them (terminal log, JSON encoder reading `\\u001b`, structured-log parsers) AND bloated diagnostic message size for adversarial multi-megabyte selectors. New helper `SanitizeForDiagnosticMessage(text, maxLength)`:

- Strips C0 (`0x00..0x1F`) + DEL (`0x7F`) + C1 (`0x80..0x9F`); replaces stripped chars with **U+FFFD replacement** so the redaction is observable to a reader (vs silently dropped).
- Truncates at `maxLength` chars (80 for selector, 200 for reason) + appends an **ellipsis marker** when truncated.

Both the raw selector AND the exception reason are sanitized — the reason field can embed selector fragments verbatim from the parser's error path.

**C-3 [CRITICAL] — Color alpha component range validation**

Per CSS Color L4 §4.2.1 the alpha component is **range-validated** at computed-value time (out-of-range → `IACVT` → declaration discarded), NOT silently clamped to [0, 1]. RGB channels themselves DO legitimately clamp per §4.2 — but alpha is the explicit exception. Cycle-1's `TryParseAlphaComponent` was clamping silently in both forms:

- Modern slash syntax: `rgb(255 0 0 / 2.0)` → resolved to opaque red instead of emitting `CSS-PROPERTY-VALUE-INVALID-001`.
- Legacy comma form: `rgba(255, 0, 0, 1.5)` → same.
- Percentage form: `/ 200%` → same.

Fix: `pct < 0 || pct > 100` reject path for percentage alpha; `n < 0 || n > 1` reject path for number alpha. Boundary values (`0`, `1`, `0%`, `100%`) remain accepted. The clamping path was the right call for RGB channels (preserved) but the wrong call for alpha (now strict).

### Tests added

- [`tests/NetPdf.UnitTests/Phase2/Phase2DeepReviewCriticalFixesTests.cs`](tests/NetPdf.UnitTests/Phase2/Phase2DeepReviewCriticalFixesTests.cs) — 10 facts:
  - C-1: `:has()` 200-deep throws (strict context — guard manifests as exception); `:not()` 200-deep throws; reasonable 5-level nesting still compiles.
  - C-2: ANSI escape sequences + BEL + NUL stripped from diagnostic; multi-megabyte selector truncated with ellipsis; total diagnostic message bounded.
  - C-3: alpha number > 1 rejected; alpha number < 0 rejected; alpha percentage > 100 rejected; legacy `rgba(...)` alpha > 1 rejected; boundary values 0/1/0%/100% all accepted.

### Test deltas

- **3239 unit / 3365 solution-wide passing** (+10 over the prior 3229/3355 baseline).
- AOT/JIT byte-parity gate green.

## Test plan

- [x] `dotnet test NetPdf.slnx -c Release` — 3239 unit / 3365 solution-wide passing
- [x] `./scripts/aot-parity.sh` — AOT/JIT byte-parity green
- [x] All 10 critical-fix facts pass with descriptive failure messages on regression
- [x] Existing tests stay green (no regression in CascadeResolverTests, ColorResolverTests, SelectorCompilerTests)
- [ ] Roland review cycle on this PR
- [ ] Address Copilot findings if any
- [ ] Merge after [PR #13](https://github.com/raroche/NetPdf/pull/13) (or rebase to main if #13 lands first)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: tests/NetPdf.UnitTests/Phase2/Phase2DeepReviewCriticalFixesTests.cs [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+    }
+
+    private static AngleSharp.Dom.IDocument ParseDoc(string html)
+    {
+        var config = Configuration.Default
+            .WithCss()
+            .WithDefaultLoader(new LoaderOptions { IsResourceLoadingEnabled = false });
+        var ctx = BrowsingContext.New(config);
+        return ctx.OpenAsync(req => req.Content(html).Address("about:blank")).Result;
+    }
```
[the remark this answers is not in the document]
Comment (raroche) [reply tests/NetPdf.UnitTests/Phase2/Phase2DeepReviewCriticalFixesTests.cs]: Addressed in 47612a5 — `ParseDoc` is now `async ParseDocAsync` returning `Task<IDocument>`; both C-2 tests are now `async Task` and await the open. Matches the async pattern used in Phase2EndToEndTests + CssParserAdapterPreprocessTests.

File: tests/NetPdf.UnitTests/Phase2/Phase2DeepReviewCriticalFixesTests.cs [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+        var media = NetPdf.Css.Cascade.CssMediaContext.DefaultPrint;
+        var sink = new CapturingSink();
+        _ = NetPdf.Css.Cascade.CascadeResolver.Resolve(doc, [sheet], media, sink);
+
+        var dx = Assert.Single(sink.Diagnostics);
+        Assert.Contains("…", dx.Message);
+        // Message length is bounded — message format is roughly
+        // 'Invalid selector "<truncated 80 chars>…" — <reason>. Rule skipped.'
+        // so total stays under ~250 chars.
+        Assert.True(dx.Message.Length < 300, $"diagnostic message too long: {dx.Message.Length} chars");
```
[the remark this answers is not in the document]
Comment (raroche) [reply tests/NetPdf.UnitTests/Phase2/Phase2DeepReviewCriticalFixesTests.cs]: Addressed in 47612a5 — relaxed the bound from < 300 to < 400 with a comment documenting the worst-case math ("Invalid selector \"" prefix 19 + selector max 80 + "\" — " 5 + reason max 200 + ". Rule skipped." 15 ≈ 320 worst case). The assertion's purpose is catching unbounded bloat (a 10_000-char selector reaching the message verbatim would explode past 10_000), not pinning an exact figure — comment now reflects that intent.

File: tests/NetPdf.UnitTests/Phase2/Phase2DeepReviewCriticalFixesTests.cs [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+            new CssSelector(hostile),
+            System.Collections.Immutable.ImmutableArray<CssDeclaration>.Empty,
+            CssSourceLocation.Unknown);
+        var sink = new CapturingSink();
+
+        // Reach the diagnostic emission via a dedicated probe — the
+        // CompileSelectorWithDiagnostics path is internal but called from
+        // CascadeResolver.CollectRules. We compile a stylesheet with a single
+        // rule whose selector is hostile + no rules to avoid a separate
+        // failure.
```
[the remark this answers is not in the document]
Comment (raroche) [reply tests/NetPdf.UnitTests/Phase2/Phase2DeepReviewCriticalFixesTests.cs]: Addressed in 47612a5 — comment now reads "hostile selector + empty declarations" which matches what the test actually builds (a stylesheet with a single rule, hostile selector, empty Declarations array).
